flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBuffer.java

```diff
@@ -0,0 +1,494 @@
+	 * @throws Exception Exceptions from async operation.
+	 */
+	public AsyncCollector<IN, OUT> add(Watermark watermark) throws Exception {
+		return processMark(watermark);
+	}
+
+	/**
+	 * Add a {@link LatencyMarker} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will be blocked here.
+	 *
+	 * @param latencyMarker LatencyMarker
+	 * @return AsyncCollector
+	 * @throws Exception Exceptions from async operation.
+	 */
+	public AsyncCollector<IN, OUT> add(LatencyMarker latencyMarker) throws Exception {
+		return processMark(latencyMarker);
+	}
+
+	private AsyncCollector<IN, OUT> processMark(StreamElement mark) throws Exception {
+		try {
+			lock.lock();
+
+			notifyCheckpointDone();
+
+			while (queue.size() >= bufferSize)
+				notFull.await();
+
+			if (error != null) {
+				throw error;
+			}
+
+			AsyncCollector<IN, OUT> collector = new AsyncCollector(this, true);
+
+			collectorToQueue.put(collector, queue.add(collector));
+			collectorToStreamElement.put(collector, mark);
+
+			// signal emitter thread that current collector is ready
+			mark(collector);
+
+			return collector;
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	/**
+	 * Notify the Emitter Thread that an AsyncCollector has completed.
+	 *
+	 * @param collector Completed AsyncCollector
+	 */
+	void mark(AsyncCollector<IN, OUT> collector) {
+		try {
+			lock.lock();
+
+			if (mode == AsyncDataStream.OutputMode.UNORDERED) {
+				finishedCollectors.add(collector);
+			}
+
+			taskDone.signal();
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	/**
+	 * Caller will wait here if buffer is not empty, meaning that not all async i/o tasks have returned yet.
+	 *
+	 * @throws Exception InterruptedException or Exceptions from AsyncCollector.
+	 */
+	void waitEmpty() throws Exception {
+		try {
+			lock.lock();
+
+			notifyCheckpointDone();
+
+			while (queue.size() != 0)
+				isEmpty.await();
+
+			if (error != null) {
+				throw error;
+			}
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	public void startEmitterThread() {
+		this.emitThread.start();
+	}
+
+	public void stopEmitterThread() {
+		emitter.stop();
+
+		emitThread.interrupt();
+	}
+
+	/**
+	 * Get all StreamElements in the AsyncCollector queue.
+	 * <p>
+	 * Emitter Thread can not output records and will wait for a while due to isCheckpointing flag
+	 * until doing checkpoint has done.
+	 *
+	 * @return A List containing StreamElements.
+	 */
+	public List<StreamElement> getStreamElementsInBuffer() {
+		try {
+			lock.lock();
+
+			// stop emitter thread
+			isCheckpointing = true;
+
+			List<StreamElement> ret = new ArrayList<>();
+			for (int i = 0; i < queue.size(); ++i) {
+				AsyncCollector<IN, OUT> collector = queue.get(i);
+				ret.add(collectorToStreamElement.get(collector));
+			}
+
+			return ret;
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	public void setOutput(TimestampedCollector<OUT> collector, Output<StreamRecord<OUT>> output) {
```

Comment:
Can't we pass the `collector` and the `output` to the constructor of the `AsyncCollectorBuffer`? Then we could make these fields final as well.